graphql_compiler/schema_generation/schema_graph.py

```diff
@@ -181,6 +131,18 @@ def freeze(self):
         """Make the SchemaElement's connections immutable."""
         self.in_connections = frozenset(self.in_connections)
         self.out_connections = frozenset(self.out_connections)
+        self._validate_endpoints()
+
+    def _validate_endpoints(self):
+        if self._kind == SchemaElement.ELEMENT_KIND_EDGE:
+            if not self.abstract:
```

Comment:
How about switching the order of the `if` and `else` here, and avoiding the `not` on this line? Too many "not"s make the code harder to read, and removing this one seems to be straightforward.